oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,358 @@
+    }
+    return AwsDates.generateXAmzDate();
+  }
+
+  private byte[] sign(byte[] key, byte[] value) {
+    try {
+      String algorithm = "HmacSHA256";
+      Mac mac = Mac.getInstance(algorithm);
+      mac.init(new SecretKeySpec(key, algorithm));
+      return mac.doFinal(value);
+    } catch (NoSuchAlgorithmException | InvalidKeyException ex) {
+      throw new SigningException("Failed to calculate the AWS V4 Signature.", ex);
+    }
+  }
+
+  static Builder newBuilder(
+      AwsSecurityCredentials awsSecurityCredentials, String httpMethod, String url, String region) {
+    return new Builder(awsSecurityCredentials, httpMethod, url, region);
+  }
+
+  static class Builder {
+
+    private AwsSecurityCredentials awsSecurityCredentials;
+    private String httpMethod;
+    private String url;
+    private String region;
+
+    @Nullable private String requestPayload;
+    @Nullable private Map<String, String> additionalHeaders;
+
+    private Builder(
+        AwsSecurityCredentials awsSecurityCredentials,
+        String httpMethod,
+        String url,
+        String region) {
+      this.awsSecurityCredentials = awsSecurityCredentials;
+      this.httpMethod = httpMethod;
+      this.url = url;
+      this.region = region;
+    }
+
+    Builder setRequestPayload(String requestPayload) {
+      this.requestPayload = requestPayload;
+      return this;
+    }
+
+    Builder setAdditionalHeaders(Map<String, String> additionalHeaders) {
+      this.additionalHeaders = additionalHeaders;
+      return this;
+    }
+
+    AwsRequestSigner build() {
+      return new AwsRequestSigner(
+          awsSecurityCredentials, httpMethod, url, region, requestPayload, additionalHeaders);
+    }
+  }
+
+  static final class AwsDates {
+
+    private static final String X_AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String CUSTOM_DATE_FORMAT = "E, dd MMM yyyy HH:mm:ss z";
+
+    private String originalDate;
+    private String amzDate;
+    private String formattedDate;
+
+    private AwsDates(String originalDate, String amzDate, String formattedDate) {
+      this.originalDate = checkNotNull(originalDate);
+      this.amzDate = checkNotNull(amzDate);
+      this.formattedDate = checkNotNull(formattedDate);
+    }
+
+    static AwsDates fromXAmzDate(String xAmzDate) {
+      return new AwsDates(xAmzDate, xAmzDate, xAmzDate.substring(0, 8));
+    }
+
+    static AwsDates fromDateHeader(String date) {
+      DateFormat dateFormat = new SimpleDateFormat(X_AMZ_DATE_FORMAT);
+      dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+      try {
+        Date inputDate = new SimpleDateFormat(CUSTOM_DATE_FORMAT).parse(date);
+        String xAmzDate = dateFormat.format(inputDate);
+        return new AwsDates(date, xAmzDate, xAmzDate.substring(0, 8));
+      } catch (ParseException e) {
+        throw new IllegalArgumentException("Invalid date provided: " + date, e);
```

Comment:
this probably shouldn't be a runtime exception; better to bubble the ParseException

Reply:
Should I bubble this up to refreshAccessToken(), and then wrap & throw it as an IOException?

Reply:
What does this exception mean? It's probably not an I?O error if the data can be read, but if the exception is caused by external input it should be checked.

Choose the exception type that fits the problem. Don't force it into an existing type that means something different.

Reply:
It's not an IO error, but I can't just bubble up the parse exception. This utility is used in refreshAccessToken - having that throw new exceptions is a breaking change (?). 

Why do you think throwing an IllegalArgumentException is not the way to go here?

Reply:
Where does the date come from? IllegalArgumentException means the calling method has a bug and should not have passed this value to this method. That doesn't seem to be the case here.

Reply:
The date can be specified as an additional header. We use the x-amz-date header, so this code path is not used in our flow. I added this because the AWS implementation supports it, and it allows us to use their tests.